### ISSUE TITLE
id2typ.pckl.  and entity_type.npy

### ISSUE THREAD

**Issue by Fatimah-Alotaibi** [zhiweihu1103/QE-TEMP#5]
Hi

Thank you so much for sharing your excellent work.

I'm working on a different dataset, and I would like to create id2typ.pckl and entity_type.npy files.

My assumption for id2typ.pckl  is just random id's for all entities and relations types. is that right?

Also, what do the lists in entity_type.npy represents?

I really appreciate any help you can provide.

**Comment by zhiweihu1103** [comment]
The type knowledge of the entity comes with the graph, but the relationship itself has no type knowledge. We use the type of the entity and the entity-relation graph to constructed the type-relation graph.

For each line in entity_type.npy represents the corresponding the type knowledge of the entity corresponding to the id of the line number.

**Comment by Fatimah-Alotaibi** [comment]
I got you thank you so much.

**Comment by Fatimah-Alotaibi** [comment]
I've got another question, I run the code and the training is very slow, is this normal behavior?

**Comment by zhiweihu1103** [comment]
Can you give your running log?

**Comment by Fatimah-Alotaibi** [comment]
2023-02-21 16:00:35 INFO     ---------------------------------------------------------------------------------------------
2023-02-21 16:00:35 INFO     Geo: beta
2023-02-21 16:00:35 INFO     Data Path: data/OOD
2023-02-21 16:00:35 INFO     #entity: 30757
2023-02-21 16:00:35 INFO     #relation: 5932
2023-02-21 16:00:35 INFO     #max steps: 450001
2023-02-21 16:00:35 INFO     Evaluate unoins using: DNF
2023-02-21 16:00:35 INFO     loading data
2023-02-21 16:00:35 INFO     Training info:
2023-02-21 16:00:35 INFO     1p: 24054
2023-02-21 16:00:35 INFO     2p: 2652
2023-02-21 16:00:35 INFO     3p: 279
2023-02-21 16:00:35 INFO     2i: 4197
2023-02-21 16:00:35 INFO     3i: 128
2023-02-21 16:00:35 INFO     ip: 238
2023-02-21 16:00:35 INFO     pi: 717
2023-02-21 16:00:35 INFO     Validation info:
2023-02-21 16:00:35 INFO     1p: 1875
2023-02-21 16:00:35 INFO     2p: 86
2023-02-21 16:00:35 INFO     3p: 1
2023-02-21 16:00:35 INFO     2i: 183
2023-02-21 16:00:35 INFO     3i: 11
2023-02-21 16:00:35 INFO     ip: 20
2023-02-21 16:00:35 INFO     pi: 18
2023-02-21 16:00:35 INFO     Test info:
2023-02-21 16:00:35 INFO     1p: 1875
2023-02-21 16:00:35 INFO     2p: 86
2023-02-21 16:00:35 INFO     3p: 1
2023-02-21 16:00:35 INFO     2i: 183
2023-02-21 16:00:35 INFO     3i: 11
2023-02-21 16:00:35 INFO     ip: 20
2023-02-21 16:00:35 INFO     pi: 18
2023-02-21 16:00:36 INFO     Model Parameter Configuration:
2023-02-21 16:00:36 INFO     Parameter gamma: torch.Size([1]), require_grad = False
2023-02-21 16:00:36 INFO     Parameter embedding_range: torch.Size([1]), require_grad = False
2023-02-21 16:00:36 INFO     Parameter entity_embedding: torch.Size([30757, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter relation_embedding: torch.Size([5932, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter type_embedding: torch.Size([1385, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.nonlinear.0.weight: torch.Size([32, 32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.nonlinear.0.bias: torch.Size([32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.nonlinear.1.weight: torch.Size([32, 32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.nonlinear.1.bias: torch.Size([32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.gate.0.weight: torch.Size([32, 32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.gate.0.bias: torch.Size([32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.gate.1.weight: torch.Size([32, 32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.gate.1.bias: torch.Size([32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.final_linear_layer.weight: torch.Size([1, 32]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.proj_layer.final_linear_layer.bias: torch.Size([1]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.layer.weight: torch.Size([1600, 3200]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter ent_neighbor_type_agg.layer.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter rel_neighbor_type_agg.linear.weight: torch.Size([800, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter rel_neighbor_type_agg.linear.bias: torch.Size([800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter rel_neighbor_type_agg.linear2.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter rel_neighbor_type_agg.linear2.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter match.map_linear.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter match.map_linear.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter match.trans_linear.weight: torch.Size([800, 800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter match.trans_linear.bias: torch.Size([800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter center_net.layer1.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter center_net.layer1.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter center_net.layer2.weight: torch.Size([800, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter center_net.layer2.bias: torch.Size([800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter relation_center_net.layer1.weight: torch.Size([800, 800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter relation_center_net.layer1.bias: torch.Size([800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter relation_center_net.layer2.weight: torch.Size([800, 800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter relation_center_net.layer2.bias: torch.Size([800]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer1.weight: torch.Size([1600, 3200]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer1.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer0.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer0.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer2.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_net.layer2.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer1.weight: torch.Size([1600, 3200]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer1.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer0.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer0.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer2.weight: torch.Size([1600, 1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter projection_without_net.layer2.bias: torch.Size([1600]), require_grad = True
2023-02-21 16:00:36 INFO     Parameter Number: 98702657
2023-02-21 16:00:36 INFO     Ramdomly Initializing beta Model...
2023-02-21 16:00:36 INFO     beta mode = (1600,2)
2023-02-21 16:00:36 INFO     tasks = 1p.2p.3p.2i.3i.ip.pi.2u.up
2023-02-21 16:00:36 INFO     init_step = 0
2023-02-21 16:00:36 INFO     Start Training...
2023-02-21 16:00:36 INFO     learning_rate = 0
2023-02-21 16:00:36 INFO     batch_size = 512
2023-02-21 16:00:36 INFO     hidden_dim = 800
2023-02-21 16:00:36 INFO     gamma = 24.000000
2023-02-21 16:00:53 INFO     Training average positive_sample_loss at step 0: 135.953064
2023-02-21 16:00:53 INFO     Training average negative_sample_loss at step 0: 0.000000
2023-02-21 16:00:53 INFO     Training average loss at step 0: 67.976532


 I used my own data with BetaE and did not take that long. however, here it takes around 15 minutes to calculate the loss for each training step.

**Comment by zhiweihu1103** [comment]
Did you use the --cuda parameter? I see that your dataset is very small, and the running time should not be 15 minutes. I run it on the dataset provided by BetaE paper, and it only takes more than ten seconds for one step.

**Comment by Fatimah-Alotaibi** [comment]
I really appreciate your quick reply. Yes, I did.  I got this warning, it might be related to this prolonged training.
UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  /opt/conda/conda-bld/pytorch_1640811805959/work/torch/csrc/utils/tensor_new.cpp:201.)
  self.entity2type = torch.tensor(entity2type) .. have you encounter this before?

**Comment by zhiweihu1103** [comment]
No, I didn't encounter this warning, I think it may be caused by the following two reasons: 
1. we use different python or pytorch versions, and the version I use can be seen in ReadMe;
2. At the following line of code
https://github.com/zhiweihu1103/QE-TEMP/blob/5dd0e5e0e48d4e7ab8b1a19c1de4623e610908f5/models.py#L399
the data format in your *entity2type* may be of *numpy.ndarrays* type. It is recommended to change it to: 
**self.entity2type = torch.tensor(np.array(entity2type)).cuda()**
If the above operations do not work, I suggest you test the speed on my dataset to rule out whether it is caused by the dataset.
